data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java

```diff
@@ -339,139 +396,57 @@ public InputStream executeStreamingQuery(String database, String command) throws
-        }
-        // trace httpCall
-        return MonitoredActivity.invoke(
-                (SupplierTwoExceptions<InputStream, DataServiceException, DataClientException>) () -> HttpPostUtils.postToStreamingOutput(httpClient,
-                        clusterEndpoint,
-                        jsonPayload, timeoutMs + CLIENT_SERVER_DELTA_IN_MILLISECS, headers),
-                "ClientImpl.executeStreamingQuery", updateAndGetExecuteTracingAttributes(database, properties));
+        KustoRequest kr = new KustoRequest(command, database, properties);
+        return executeStreamingQuery(kr);
     }
 
-    private long determineTimeout(ClientRequestProperties properties, CommandType commandType, String clusterUrl) throws DataClientException {
-        Long timeoutMs;
-        try {
-            timeoutMs = properties == null ? null : properties.getTimeoutInMilliSec();
-        } catch (ParseException e) {
-            throw new DataClientException(clusterUrl, "Failed to parse timeout from ClientRequestProperties");
-        }
-        if (timeoutMs == null) {
-            if (commandType == CommandType.ADMIN_COMMAND) {
-                timeoutMs = COMMAND_TIMEOUT_IN_MILLISECS;
-            } else {
-                timeoutMs = QUERY_TIMEOUT_IN_MILLISECS;
-            }
-        }
-        return timeoutMs;
-    }
-
-    private CommandType determineCommandType(String command) {
-        if (command.startsWith(ADMIN_COMMANDS_PREFIX)) {
-            return CommandType.ADMIN_COMMAND;
-        }
-        return CommandType.QUERY;
-    }
-
-    private Map<String, String> generateIngestAndCommandHeaders(ClientRequestProperties properties,
-            String clientRequestIdPrefix,
-            String activityTypeSuffix)
-            throws DataServiceException, DataClientException {
-
-        Map<String, String> headers = extractTracingHeaders(properties);
-
-        if (aadAuthenticationHelper != null) {
-            headers.put(HttpHeaders.AUTHORIZATION, String.format("Bearer %s", aadAuthenticationHelper.acquireAccessToken()));
-        }
-
-        String clientRequestId;
-        if (properties != null && StringUtils.isNotBlank(properties.getClientRequestId())) {
-            clientRequestId = properties.getClientRequestId();
-        } else {
-            clientRequestId = String.format("%s;%s", clientRequestIdPrefix, UUID.randomUUID());
-        }
-
-        headers.put("x-ms-client-request-id", clientRequestId);
-
-        headers.put("Connection", "Keep-Alive");
-
-        UUID activityId = UUID.randomUUID();
-        String activityContext = String.format("%s%s/%s, ActivityId=%s, ParentId=%s, ClientRequestId=%s", JAVA_INGEST_ACTIVITY_TYPE_PREFIX, activityTypeSuffix,
-                activityId, activityId, activityId, clientRequestId);
-        headers.put("x-ms-activitycontext", activityContext);
+    public InputStream executeStreamingQuery(@NotNull KustoRequest kr) throws DataServiceException, DataClientException {
```

Comment:
It makes sense to have the request object because the object encapsulates all of the necessary state which is input for the executeStreamingQuery method. This means that if the input state needs to change in the future, the method signature does not, and therefore changes can be made without impacting users of the existing API or adding any additional method chaining. Instead you just add to your object whatever additional state you may need, allowing for the KustoRequest to be an abstract concept, and then you can easily use BiConsumers and functional paradigms like what you find in Project Reactor to write reactive code, which is designed to work with callbacks that take a singular input and provide a singular output.

Reply:
I did just change this to private though